### PR TITLE
[rhi] Update `create_pipeline` API and add support of VkPipelineCache

### PR DESCRIPTION
**Pull request by bobcao3**
Issue: #6832

### Brief Summary

The `create_pipeline` API has been updated to new standards.

In addition, we now added (optional) support for backend caches such as `VkPipelineCache`. This makes running cached taichi programs even faster.

### CODE REVIEW

File: taichi/rhi/device.h [hunk 1]
```diff
@@ -543,6 +553,28 @@ class TI_DLL_EXPORT Stream {
...
+   * - Can return `nullptr` if cache is invalid or empty.
+   */
+  virtual void *data() noexcept {
+    return nullptr;
+  }
+
+  /**
+   * Get the size of the cache (in bytes).
+   */
+  virtual size_t size() noexcept {
```
Comment (k-ye) [inline taichi/rhi/device.h:571]: super nit: const?

**Review by k-ye — APPROVED** [review]
LGTM!

File: taichi/rhi/device.h [hunk 2]
```diff
@@ -570,9 +602,64 @@ class TI_DLL_EXPORT Device {
...
+   */
+  virtual RhiResult create_pipeline_cache(
+      PipelineCache **out_cache,
+      size_t initial_size = 0,
+      const void *initial_data = nullptr) noexcept {
+    *out_cache = nullptr;
+    return RhiResult::not_supported;
+  }
+
+  inline std::pair<UPipelineCache, RhiResult> create_pipeline_cache_unique(
```
[the remark this answers is not in the document]
Comment (bobcao3) [reply taichi/rhi/device.h:626]: We can in device.h if it doesn't manifest into the binary. The ABI is still protected

[the remark this answers is not in the document]
Comment (bobcao3) [reply taichi/rhi/device.h:626]: Unless I'm mistaken and this went into the symbol table afterall...

[the remark this answers is not in the document]
Comment (k-ye) [reply taichi/rhi/device.h:626]: cc @jim19930609 ...

**Comment by taichi-gardener** [comment]
Rebase failed, please resolve the conflicts manually.

**Pull request merged**